src/LndClient/Data/LndEnv.hs

```diff
@@ -1,58 +1,133 @@
 import Data.Text.Lazy as LT
 import Data.X509
 import Env
 import LndClient.Class
 import LndClient.Data.Type
-import LndClient.Import.External
+import LndClient.Import.External as Ex
 import LndClient.Util as U
 import Network.GRPC.HighLevel.Generated
 import Network.GRPC.LowLevel.Client
+import Text.Read
 
-newtype LndWalletPassword = LndWalletPassword ByteString
-  deriving (PersistField, PersistFieldSql, Eq)
+newtype LndWalletPassword = LndWalletPassword Text
+  deriving (PersistField, PersistFieldSql, Eq, FromJSON, IsString, Show)
 
 newtype LndTlsCert = LndTlsCert ByteString
-  deriving (PersistField, PersistFieldSql, Eq)
+  deriving (PersistField, PersistFieldSql, Eq, Show)
 
-newtype LndHexMacaroon = LndHexMacaroon ByteString
-  deriving (PersistField, PersistFieldSql, Eq)
+createLndTlsCert :: ByteString -> Either LndError LndTlsCert
+createLndTlsCert bs = do
+  pemsM <- first (LndEnvError . LT.pack) $ Pem.pemParseBS bs
+  pem <-
+    note (LndEnvError $ LT.pack "No pem found") $ safeHead pemsM
+  bimap
+    (LndEnvError . LT.pack . ("Certificate is not valid: " <>))
+    (const $ LndTlsCert bs)
+    (decodeSignedCertificate $ Pem.pemContent pem)
 
-newtype LndHost = LndHost ByteString
-  deriving (PersistField, PersistFieldSql, Eq)
+instance Read LndTlsCert where
+  readsPrec _ x = do
+    case createLndTlsCert $ C8.pack x of
+      Right r -> [(r, "")]
+      Left _ -> case reads x of
+        [(r, "")] -> [(r, "")]
+        _ -> []
 
-newtype LndPort = LndPort Word32
-  deriving (PersistField, PersistFieldSql, Eq)
+instance FromJSON LndTlsCert where
+  parseJSON x =
+    case x of
+      A.String s ->
+        case createLndTlsCert $ C8.pack $ T.unpack s of
+          Right cert -> return cert
+          Left e -> failure e
+      e -> failure e
+    where
+      failure err = fail $ "Json certificate parsing error: " <> " " <> show err
 
-instance ToGrpc LndWalletPassword ByteString where
+instance ToGrpc LndTlsCert ByteString where
+  toGrpc = Right . coerce
+
+newtype LndHexMacaroon = LndHexMacaroon Text
+  deriving (PersistField, PersistFieldSql, Eq, FromJSON, IsString, Show)
+
+newtype LndHost = LndHost Text
+  deriving (PersistField, PersistFieldSql, Eq, FromJSON, IsString, Show)
+
+newtype LndPort = LndPort Int
+  deriving (PersistField, PersistFieldSql, Eq, Show)
+
+createLndPort :: Word32 -> Either LndError LndPort
+createLndPort p = do
+  let maybePort :: Maybe Int = U.safeFromIntegral p
+  maybeToRight (LndEnvError "Wrong port") $ LndPort <$> maybePort
+
+instance Read LndPort where
+  readsPrec _ x = do
+    let mX :: Maybe Word32 = readMaybe x
+    let eX :: Either LndError Word32 = maybeToRight (LndEnvError "Port parsing error") mX
+    case createLndPort =<< eX of
+      Right r -> [(r, "")]
+      Left _ -> case reads x of
+        [(r, "")] -> [(r, "")]
+        _ -> []
+
+instance FromJSON LndPort where
+  parseJSON x =
+    case x of
+      A.Number s -> do
+        let maybeInt :: Maybe Int = toBoundedInteger s
```

Comment:
seems like `toBoundedInteger` function is polymorphic, so you can obtain there `Maybe Word32` so, don't extra `safeFromIntegral` computation in next line

Reply:
Okay)